alica_viewer/include/alica_viewer/alica_plan_parser.h

```diff
@@ -0,0 +1,277 @@
+    std::unordered_set<std::shared_ptr<CombinedSimplePlanTree>> _children;
+    AgentGrp _robotIds;
+};
+
+struct StateHash
+{
+    std::size_t operator()(const State* const obj) const { return std::hash<int64_t>{}(obj->getId()); }
+};
+
+struct StateEqualsComparator
+{
+    bool operator()(const State* a, const State* b) const { return a->getId() == b->getId(); }
+};
+
+typedef std::unordered_map<const State*, std::shared_ptr<CombinedSimplePlanTree>, StateHash, StateEqualsComparator> CombinedSimplePlanTreeMap;
+
+class AlicaPlan
+{
+  public:
+    AlicaPlan(int argc, char* argv[])
+        : _planParser(&_planRepository)
+        , _agentIDManager(new supplementary::AgentIDFactory())
+        , _rootState(new State(-1))
+    {
+        if (argc < 2) {
+            std::cout << "Usage: Base -m [Masterplan] -rd [rolesetdir] -r [roleset]" << std::endl;
+            exit(1);
+        }
+
+        std::string masterPlanName = "";
+        std::string roleSetName = "";
+        std::string roleSetDir = "";
+
+        for (int i = 1; i < argc; ++i) {
+            if (std::string(argv[i]) == "-m" || std::string(argv[i]) == "-masterplan") {
+                masterPlanName = argv[i + 1];
+                ++i;
+            }
+            if (std::string(argv[i]) == "-rd" || std::string(argv[i]) == "-rolesetdir") {
+                roleSetDir = argv[i + 1];
+                ++i;
+            }
+            if (std::string(argv[i]) == "-r" || std::string(argv[i]) == "-roleset") {
+                roleSetName = argv[i + 1];
+                ++i;
+            }
+        }
+
+        if (masterPlanName.size() == 0 || roleSetDir.size() == 0) {
+            std::cout << "Usage: Base -m [Masterplan] -rd [rolesetdir] -r [roleset]" << std::endl;
+            exit(1);
+        }
+        std::cout << "Masterplan is: " << masterPlanName << std::endl;
+        std::cout << "Rolset Directory is: " << roleSetDir << std::endl;
+        std::cout << "Rolset is: " << roleSetName << std::endl;
+
+        _planParser.parsePlanTree(masterPlanName);
+
+        ///@todo remove this once AgentID becomes independent of Globals.conf
+        std::shared_ptr<std::vector<std::string>> agentNames = (*supplementary::SystemConfig::getInstance())["Globals"]->getSections("Globals.Team", NULL);
+        for (const std::string& agentName : *agentNames) {
+            int id = (*supplementary::SystemConfig::getInstance())["Globals"]->tryGet<int>(-1, "Globals", "Team", agentName.c_str(), "ID", NULL);
+            _agentInfos.emplace(_agentIDManager.getID(id), AgentInfo(id, agentName));
+        }
+    }
+
+    ~AlicaPlan() { delete _rootState; }
+
+    void addSptToCspt(std::shared_ptr<CombinedSimplePlanTree> parent, const std::shared_ptr<SimplePlanTree> spt)
+    {
+        const State* state = spt->getState();
+        std::shared_ptr<CombinedSimplePlanTree> cur;
+        CombinedSimplePlanTreeMap::iterator csptEntry = _combinedSimplePlanTree.find(state);
```

Comment:
As we talked about, this map is not really a good way to merge the trees, as the same ids may occur in different contexts. Merging from root to leaf is better.